### PR TITLE
Now all in one Driver files for WHB04B-6 and update status for main

### PR DESCRIPTION
**Pull request by alkabal**
I have now tested the driver as much as i can.

All seem to work great regading original function and i have enhanced some of them.

### Button enhancement : 
Single function button Macro10 are now used as a standard button for toggle Dro Absolue/Relative and can also be used as Macro14 with push and hold the Fn button like the other button with macro.

Single function button Continuous (Mpg mode) have now a macro function that toggle the Mist state regading the internal status (main.py regarding smoothie $G switch states)

Single function button Continuous (Step mode) have now a macro function that toggle the Mist state regading the internal status (main.py regarding smoothie $G switch states)

Button Reset/Stop/Play can also be used as Macro11/12/13 if needed.

### Function enhancement : 
_At startup or after reset whb04b-6/Smoopi or each time you set Lead = the Mpg/Step/Con mode is allways revert to Con mode._

Con mode = feedrate override using the wheel with selectable range from the rotary knob
Mpg mode = as usual
Step mode = as usual
Lead position = move from Mpg wheel locked but keep other button working. (maybee i think is possible to add later a Spindle override from wheel if this can be usefull ?)

### The readme need more details

Br

### CODE REVIEW

**Comment by alkabal** [comment]
Thanks for keep this PR open ! i have try to do better, but i see they are too much commit i appologize.

Br

**Comment by alkabal** [comment]
For the question i think this parts of the code does not work great, inside your driver you have commented "self.s_ovr = i"

```
    def update_fro(self, i, v):
        self.s_ovr = i                      # I think does not work
        self.f_ovr = v                      # update pendant feed override from smoothie value if changed by smoothie
        self.setovr(self.f_ovr, self.s_ovr)
```

File: main.py [hunk 1]
```diff
@@ -617,7 +619,7 @@ def stream_finished(self, ok):
         self.display(">>> Elapsed time: {}".format(et))
         self.eta = '--:--:--'
 
-    def upload_gcode(self):
+     def upload_gcode(self):
```
Comment (wolfmanjm) [inline main.py]: indent error.

**Comment by alkabal** [comment]
Hi

Regarding to the patch done in jogrose for use the B axis in CNC mode i see the missing ABC value on the left bottom on the screen using the jog panel. (also missing C axis but i think all of this need to be conditional like the Dro panel displaying N/A for C axis)

Also the Mpg wheel does not allow use ABC.

Br

**Comment by wolfmanjm** [comment]
Very few (if Any) have B and C axis, so this is not a priority for me.

**Comment by wolfmanjm** [comment]
Hi, I am teting this with a WHB04B-4 wired version (not wireless). It does not seem to work reliably at all, I am not sure if the code is just buggy, or the unit I have is using a slightly different protocol.
Main problem is that many buttons when pressed continously repeat which for the toggle buttons is an issue as you have to hit it very quickly to get what you want. It does not ever seem to workj in MPG mode it just says illegal axis F.
I'll see if the protocol is different and maybe will need a new module.
I have merged with the current master.

**Comment by wolfmanjm** [comment]
Ok I see the F error is becuase smoothie does not accept $J F anymore, F is now S.

**Comment by alkabal** [comment]
Hi

I think other things need review because I have not use this machine from some times 

I like to change some things like default mode mpg/con/step and review pulse management 

In the past you also ask for make less verbosity 

Br

**Comment by wolfmanjm** [comment]
FYI I've done my own version in the branch https://github.com/wolfmanjm/kivy-smoothie-host/tree/add/whb04b

This fixes a lot of issues I had with this PR.
Mainly repeating buttons (they need to be debounced), and LCD Position displays did not correctly display coordinates, so I took code from the linuxcnc port to format them correctly.

However I did not implement many of the builtin functions here as I disagreed with these being hard coded. Most of them can be implemented in the macros file. (mist and flood would need to be added like I have done the find-center or half zero)

I particularly disagree with the reset button connecting to smoothie.

The only thing I can't get to work at the moment is the display of the F override and S override they seem to always display 0 however the buttons do work and change the f override. I will consider implementing the speed overides using the lead mode and the wheel, I need to think about that though.

I have not implemented the MPG or CONT modes or the Lead as I am not really sure what they should do on smoothie. However the modes can be set and the display shows them.

This will almost certainly work with the wireless version.
It will probably work with the 6 axis version with a little bit of work.

I do need to fix the display of the ABC axis when in those modes.

Thank you for your efforts, your code did help quite a lot even though I didn't use any of it, it was useful as a reference.

**Comment by wolfmanjm** [comment]
WHB04B now implemented in master. It is a different one to this, but works well.

**Comment by wolfmanjm** [comment]
@alkabal Did you ever get the F: or S: to ever display anything other than 0? I have tried everythig I can think of and it never changes.

**Comment by alkabal** [comment]
F14D0D77-31D2-417B-BCE2-97D15BA9407D
68B01348-F47B-4FC3-B8C1-C7525E37CBA4
9157F4CE-86B1-407A-9C6D-F42E06AEE69A
D0A35F0E-32F9-4263-A7B9-2210D9C8DAC4




Yes but only in some situation using the button feed+ feed-

**Comment by alkabal** [comment]
The reset button in LinuxCNC is connected like I have done in the driver.

I don’t understand why you have some issue with debounce except the difference from usb to wireless 

I have do the best for doing something very similar to lcnc (but now my code need some change to be reflected)

Br

**Pull request closed without merging**